python/tank/util/user_settings.py

```diff
@@ -0,0 +1,246 @@
+                    **kwargs
+                )
+                instance._init_singleton()
+
+                # remember the instance so that no more are created
+                cls._instance = instance
+
+        return cls._instance
+
+    @classmethod
+    def reset_singleton(cls):
+        """
+        Resets the internal singleton instance.
+        """
+        cls._instance = None
+
+
+class UserSettings(Singleton):
+    """
+    Handles finding and loading the user settings for Toolkit.
+    """
+
+    _LOGIN = "Login"
+
+    @property
+    def default_http_proxy(self):
+        """
+        :returns: The default proxy.
+        """
+        return self._get_value("http_proxy")
+
+    def is_default_app_store_http_proxy_set(self):
+        """
+        :returns: ``True`` if ``app_store_http_proxy`` is set, ``False`` otherwise.
+        """
+        return self._is_setting_found("app_store_http_proxy")
+
+    @property
+    def default_app_store_http_proxy(self):
+        """
+        :returns: The app store specific proxy or None is not set or set to an empty value.
+            Use is_default_app_store_http_proxy_set to disambiguate.
+        """
+        # If the config parser returned a falsy value, it meant that the app_store_http_proxy
+        # setting was present but empty. We'll advertise that fact as None instead.
+        return self._get_value("app_store_http_proxy") or None
+
+    @property
+    def default_site(self):
+        """
+        :returns: The default site.
+        """
+        return self._get_value("default_site")
+
+    @property
+    def default_login(self):
+        """
+        :returns: The default login.
+        """
+        return self._get_value("default_login")
+
+    def _get_value(self, key):
+        """
+        Retrieves a value from the config.ini file. If the value is not set, returns the default.
+        Since all values are strings inside the file, you can optionally cast the data to another type.
+
+        :param default: If the value is not found, returns this default value. Defaults to None.
+
+        :returns: The appropriately type casted value if the value is found, default otherwise.
+        """
+        if not self._is_setting_found(key):
+            return None
+        else:
+            # Read the value, remove any extra whitespace. If the string is empty,
+            # or None will set value to None
+            value = os.path.expandvars(self._user_config.get(self._LOGIN, key))
+            return value.strip()
+
+    def _is_setting_found(self, key):
+        """
+        Checks if the setting is in the file.
+
+        :param key: Name of the setting within the Login section.
+
+        :returns: True if found, False otherwise.
+        """
+        if not self._user_config.has_section(self._LOGIN):
+            return False
+        elif not self._user_config.has_option(self._LOGIN, key):
+            return False
+        return True
+
+    def _init_singleton(self):
```

Comment:
nitpicky: normally, we put constructors at the top of the file - in this case, this method feels similar to the `__init__`  - so maybe it should appear first in the class?